Places survey

sebastianbarry opened the pull request.

shankari commented:
I'm glad you were able to fix that - it will unblock me as well.
However, I don't understand the change. As the comment at the top says, the 

```
        // The trip structure is different between the diary and label screens
        // one has the timestamps in properties and the other does not
        // let's support both so we can label from either screen
``` 

If you try to label the trip from the diary, for example, it **will** have `data.properties` filled out

That's why I have the line

```
        let tsObj = _state.opts.trip.data? _state.opts.trip.data.properties : _state.opts.trip;
```

From your screenshots, it looks like the trip has `data` with `features` but no properties.
Is this from the label screen or the diary screen?
I wonder if this is a regression triggered by the changes in https://github.com/e-mission/e-mission-phone/blob/master/www/js/diary/infinite_scroll_list.js#L344

- can you test with a change prior to the diary/label refactor? (e.g. `git checkout 82c997cb32b6ca23c31a75f2f78ce643951d1d45`) and confirm that the code works even without your changes?
- if so, the fix might be to change `Timeline.confirmedTrip2Geojson` to actually fill in the properties correctly instead of adding additional hacks. that would also allow us to remove the original hack above (`let tsObj = ...`)

sebastianbarry commented:
My previous screenshots were from the Label screen.

When labeling from the Diary screen, `_scope.opts.trip.properties` does in fact have `start_ts` and `end_ts`, unlike labeling from the Label screen:

<img width="1173" alt="image" src="https://user-images.githubusercontent.com/61334340/210626105-7d8839ab-190a-423b-8eef-39136c900dab.png">

I will no do the following:

> - can you test with a change prior to the diary/label refactor? (e.g. git checkout 82c997cb32b6ca23c31a75f2f78ce643951d1d45) and confirm that the code works even without your changes?
> - if so, the fix might be to change Timeline.confirmedTrip2Geojson to actually fill in the properties correctly instead of adding additional hacks. that would also allow us to remove the original hack above (let tsObj = ...)

sebastianbarry commented:
> - can you test with a change prior to the diary/label refactor? (e.g. git checkout https://github.com/e-mission/e-mission-phone/commit/82c997cb32b6ca23c31a75f2f78ce643951d1d45) and confirm that the code works even without your changes?

The code works in this way:

Old Label Screen:
* `_state.opts.trip.data.properties` is still undefined _(the same behavior as our current commit)_
* `_state.opts.trip.data?` evaluates to false, so `_state.opts.trip` is used instead of `_state.opts.trip.data.properties`
* `_state.opts.trip` contains the `start_ts` and the `stop_ts`

<img width="1175" alt="Screen Shot 2023-01-04 at 12 40 25 PM" src="https://user-images.githubusercontent.com/61334340/210636944-63dbb788-0067-461b-b9a5-c899017ce367.png">


Old Diary Screen:
* `_state.opts.trip.data.properties` is defined and contains `start_ts` and `end_ts`
* `_state.opts.trip.data?` evaluates to true, so `_state.opts.trip.data.properties` is used

<img width="1172" alt="Screen Shot 2023-01-04 at 12 35 29 PM" src="https://user-images.githubusercontent.com/61334340/210636963-812b5b65-5bb2-43dd-a9e6-077526b9ddc9.png">

-------

That being said, the solution appears to be:

1. Find wherever `_state.opts.trip.data` is defined, make sure that the `start_ts` and `end_ts` are added here.

@shankari you had mentioned:

> if so, the fix might be to change Timeline.confirmedTrip2Geojson to actually fill in the properties correctly instead of adding additional hacks. that would also allow us to remove the original hack above (let tsObj = ...)

shankari commented:
@sebastianbarry 

Given this: 

> _state.opts.trip.data? evaluates to false, so _state.opts.trip is used instead of _state.opts.trip.data.properties

This is clearly a regression caused by the changes to unify the diary and label screens.

Not sure why this is your proposed solution:

>  Find wherever _state.opts.trip.data is defined, make sure that the start_ts and end_ts are added here.

instead of my proposal to change `Timeline.confirmedTrip2Geojson`

Can you indicate why your proposed solution is better?

JGreenlee commented:
I don't know the cause of the regression, but don't we want to have a consistent structure in the trip object whichever page it is on?

sebastianbarry commented:
Testing for start_ts and end_ts:

I first labeled a trip using `ENKETO` `trip_user_input` on the label screen, and the trip was removed from the "to-label" filter correctly.
Then, I pressed the refresh 🔄 button and the trip remained NOT in the "to-label" filter, correctly
Then, I went to the diary screen to make sure that trip remained labeled in there, and here is what I found:
![Simulator Screen Shot - iPhone 11 - 2023-01-04 at 14 05 28](https://user-images.githubusercontent.com/61334340/210660653-0cdb22b5-44df-4fd1-9430-e6dfcc12544e.png)

Looks like it's working now! No band-aid fixes and the `trip.opts.data.properties.start_ts/end_ts` exist from both the diary screen, and now the label screen too!

JGreenlee commented:
<img width="392" alt="image" src="https://user-images.githubusercontent.com/15843932/210935989-c8ea5b47-fa8d-41b6-958d-c9c9061e01e6.png">

dc432d3 implements the list that displays trip- or place- additions that are coming from the server. After completing timeuse survey, forcing sync, running pipeline, and refreshing - they will show up at the bottom of the trip listitem
I need @sebastianbarry to help with the input matching part. I can cache survey results in the trip object, and I can grab trip additions from the server - but how do I integrate those together? (along with unprocessed additions)

A few things to note:
"Answered" is the label that results from `answer.js`. I think we will want to change this and I can think of 2 options:
1) compute a quick recap of the selected activity (eg. "Doing Laundry")
 - if multiple selected, use ellipses? (eg. "Doing Laundry, Cleaning ...")
2) or if we want to stay generic, simply a counter for the activities (eg. "1 Activity" , "2 Activities")

Something is definitely wrong with the timestamps. The timestamps of the trip_addition entries are always the same, even if they were changed in the survey itself. I'll debug this tomorrow
It would be helpful if the trip_addition entries could have a `display_start_time`/`display_end_time` field coming from the server

shankari commented:
Thanks for the changes, @JGreenlee
High level comments without looking at the code:

The question of how to display the result is the one that I highlighted in the Teams Chat as well. There is a actually a request from the Sydney folks as part of their design https://github.com/e-mission/e-mission-phone/pull/915#issuecomment-1358493728  It does have a quick recap of the selected activity with a `+n` if there are multiple selected
<img width="402" alt="Screen Shot 2022-12-20 at 12 17 33 PM" src="https://user-images.githubusercontent.com/2423263/208758580-584ff9a4-8965-417f-afaf-eb13ae392910.png">

From the example for the fake missing trips survey as well
https://github.com/e-mission/nrel-openpath-deploy-configs/pull/10 
it looks like `answer.js` is definitely going to be the next area we need to focus on for design

In Teams chat yesterday, I suggested that we might want to replace `answer.js` with extraction paths and a template in the survey spec so that we don't need to change code every time there is a new survey.

shankari commented:
Notes to self (and maybe to @JGreenlee and @sebastianbarry).
wrt filling in templates in `answer.js`
- there is a template literal mechanism built into javascript with `` (which is what `answer.js` currently uses) but it is not possible to pass in a string to do that (e.g. https://stackoverflow.com/a/29771751). We need to use a templating engine
- this seems fairly similar to the templating used in translation software - let's see what `angular-translate` uses and use the same
   - it looks like it (https://github.com/angular-translate/angular-translate/blob/master/bower.json) uses `messageformat` and `angular-sanitize`
- since the messages will need to be translated anyway, maybe we can just re-use the angular-translate functionality directly!

sebastianbarry commented:
After our meeting, here is our final Write-Up:

### What is the problem we are trying to solve?

_We want to add new surveys to existing button locations (i.e. adding the MissingTripSurvey in the place-notes button in replacement of TimeUseSurvey)._
As of right now, when we create a new `Survey` in one of the `nrel-openpath-dynamic-configs/configs` files to look like this: https://github.com/sebastianbarry/nrel-openpath-deploy-configs/blob/missing-trips-survey/configs/dev-emulator-program-missingtrip.nrel-op.json
```json
 "survey_info": {
      "surveys": {
        ...
        "MissingTripSurvey": {
          "formPath": "https://raw.githubusercontent.com/sebastianbarry/nrel-openpath-deploy-configs/missing-trips-survey/survey-resources/data-json/fake-missing-trip-survey-v1.json",
          "version": 9,
          "compatibleWith": 1,
          "dataKey": "manual/missingtrip-survey"
        }
      },
      "buttons": {
        "place-notes": {
          "surveyName": "MissingTripSurvey",
          "not-filled-in-label": {
            "en": "Add Missing Trip",
            "es": "Añadir viaje perdido"
          },
          "filled-in-label": {
            "en": "",
            "es": ""
          }
        }
      },
```

... then we see that the `place-notes` button appears with the correct labels. [SUCCESS](https://github.com/e-mission/nrel-openpath-deploy-configs/pull/10#issue-1522847054)
However, when we fill out the survey and attempt to click "Save" to close it out, the survey does not complete because there is no `LABEL_FUNCTIONS` answer function written in `e-mission-phone/www/js/survey/enketo/answer.js` for `the new MissingTripSurvey.

### 1st Proposed answer (not our final decision)
- We create a new `LABEL_FUNCTIONS` function for MissingTripSurvey
```javascript
const LABEL_FUNCTIONS = {
    TripConfirmSurvey: (xmlDoc) => {
      const modeStr = _getAnswerByTagName(xmlDoc, 'travel_mode');
      const purposeStr = _getAnswerByTagName(xmlDoc, 'destination_purpose');

      if (modeStr.includes('trip_not_valid') || purposeStr.includes('trip_not_valid')) {
        return 'Trip not valid';
      }

      const purposes = purposeStr.split(' ').length;
      const modes = modeStr.split(' ').length;
      return `${purposes} purpose${purposes > 1 ? 's': ''}, ${modes} mode${modes > 1 ? 's': ''}`;
    },
    UserProfileSurvey: (xmlDoc) => {
      return 'Answered';
    },
    TimeUseSurvey: (xmlDoc) => {
      return 'Answered';
    },
    MissingTripSurvey: (xmlDoc) => {
      return 'Answered';
    }
  };
```

- This requires that for every new survey we add to any `nrel-openpath-deploy-configs/configs` file, we will need to come into this `e-mission-phone/www/js/survey/enketo/answer.js` file and create a new `LABEL_FUNCTIONS` function

### 2nd Proposed answer (THIS is the one we will go with)

- We create a Generic/Templated `LABEL_FUNCTIONS` function
- ~~Any survey created that wants to return something other than `'Answered'` will have to prepend the fields they want in their survey's `answer.js` answer with `label_`:~~
<img width="1233" alt="image" src="https://user-images.githubusercontent.com/61334340/211099541-2f3df6d5-aa30-449f-b0ae-db10d986fd32.png">

  - We decided this would be too difficult because we would have to tree traverse the xmlDoc
- `nrel-openpath-deploy-configs/configs` config files will now have the fields that they want to be a part of the answer defined in the field `survey_info.surveys.[survey name].answerFields` field names separated by commas like this:
```json
"survey_info": {
      "surveys": {
        ...
        "TimeUseSurvey": {
          "formPath": "https://raw.githubusercontent.com/sebastianbarry/nrel-openpath-deploy-configs/surveys-info-and-surveys-data/survey-resources/data-json/time-use-survey-form-v9.json",
          "version": 9,
          "compatibleWith": 1,
          "answerFields": "Personal_Care_activities,Employment_related_and_Educati,Domestic_activities,Recreation_and_leisure,Voluntary_work_and_care_activi,Other",
          "dataKey": "manual/timeuse_survey"
        }
      },
      "buttons": {
        "trip-notes": {
          "surveyName": "TimeUseSurvey",
          "not-filled-in-label": {
            "en": "Add Activity",
            "es": "Añadir actividad"
          },
          "filled-in-label": {
            "en": "",
            "es": ""
          }
        },
        ...
      },
      "trip-labels": "MULTILABEL"
    },
```

- Once the answer fields have been defined in the config, there will be a Templated answer generated in `answer.js`. It will look something like this:
```javascript
  const LABEL_FUNCTIONS = {
    ...
    TemplatedSurvey: (xmlDoc, surveyName) => {
      var answer = "";
      
      // Get the dynamic config and save it to $scope.ui_config
      DynamicConfig.configReady().then(function(newConfig) {
        $scope.ui_config = newConfig;
        $scope.refreshScreen();
      });

      // Create a list of the possible answer fields that were defined in the conifg
      var possibleAnswerFieldsList = ui_config.survey_info.surveys[surveyName].answerFields.split(',');
      
      // Go through each possible answer field, and see if it exists in the response, then add the value to the answer
      possibleAnswerFieldsList.forEach(answerField => {
        if(xmlDoc.getElementsByTagName(answerField))
          answer += xmlDoc.getElementsByTagName(answerField)
      });
      
      return `${answer}`;
    }
  };
```

  - We will use a **Standardized** version of templating, which is the same as `angular-translate`: https://github.com/angular-translate/angular-translate/blob/master/bower.json

JGreenlee commented:
@shankari @sebastianbarry 

### 3rd Proposed Answer
Here is my understanding:

Clients will specify 2 fields (calling them `labelVars` and `labelTemplate` here) in the dynamic config file like so:

#### *.nrel-op.json
```json
...
 "surveys": {
   "TripConfirmSurvey": {
     "formPath": ...,
     "dataKey": ...,
     "labelVars": {
       "modes": "travel_mode",
       "purposes": "destination_purpose",
     }
     "labelTemplate": "${purposes} purpose${purposes > 1 ? 's': ''}, ${modes} mode${modes > 1 ? 's': ''}"
   }
   ...
}
...
```

This will allow a dynamic construction of a label by the person creating the server - no need to update the code.
If a client wants a label beyond the limitations of this, we write it for them

#### answer.js > LABEL_FUNCTIONS

```js

  // gather vars that will be used in the template according to the survey config
  const labelVars = {}
  for (varName in _config.labelVars) {
    const fieldName = _config.labelVars[varName];
    const fieldStr = _getAnswerByTagName(xmlDoc, fieldName);
    const fieldMatches = fieldStr.split(' ');
    labelVars[varName] = fieldMatches;
  }

  const labelTemplate = _config.labelTemplate;
  // use angular templating engine to apply labelVars to labelTemplate

  // eg. labelTemplate
  // "${purposes} purpose${purposes > 1 ? 's': ''}, ${modes} mode${modes > 1 ? 's': ''}"

  // eg. returned result
  // "1 purpose, 2 modes"
  
```

JGreenlee commented:
@shankari thanks for reviewing that. One more correction I want to make to what I proposed yesterday:

We want this approach to be better for i18n. We should be able to provide versions of `labelTemplate` in different languages, so I think it should be structured like this:

#### *.nrel-op.json
```json
...
 "surveys": {
   "TripConfirmSurvey": {
     "formPath": ...,
     "dataKey": ...,
     "labelVars": {
       "modes": "travel_mode",
       "purposes": "destination_purpose",
     }
     "labelTemplate": {
       "en": "${purposes} purpose${purposes > 1 ? 's': ''}, ${modes} mode${modes > 1 ? 's': ''}",
       "es": "${purposes} propósito${purposes > 1 ? 's': ''}, ${modes} modo${modes > 1 ? 's': ''}",
       ...
   }
   ...
}
...
```

I think I now have a solid idea of how to actually apply the template dynamically. Once @sebastianbarry can make the config changes I'll be able to see if it works!

sebastianbarry commented:
@JGreenlee I added the templates and vars to the config files!! Here is the Commit: https://github.com/e-mission/nrel-openpath-deploy-configs/pull/8/commits/685a8dabe12f2018bbc5201c5337725b77105c67

Could you give me feedback on the TimeUseSurvey template I created? I am not sure I am doing the templating right. I want it to display whatever was the second option we selected, for example:

![Simulator Screen Shot - iPhone 11 - 2023-01-07 at 10 45 37](https://user-images.githubusercontent.com/61334340/211163689-a1cd2e50-ab43-4d08-9acf-47de42e15122.png)

For this survey, "Sleeping" is what would be in the `answer.js` response

JGreenlee commented:
I have discovered that MessageFormat is more limited than previously thought.

It can interpolate values by simple substitution (and I can get this working), but it cannot access those values' properties (such as .length).

I could change `LABEL_FUNCTIONS` to just use the .length (effectively the count of selected survey inputs) but this limits us from ever displaying any info about the actual options that were selected. Too generic, I think

If we want to support both use cases (counting and listing) using templating, we need to specify more things in the config.

sebastianbarry commented:
```json
"TripConfirmSurvey": {
          "formPath": "https://raw.githubusercontent.com/sebastianbarry/nrel-openpath-deploy-configs/surveys-info-and-surveys-data/survey-resources/data-json/trip-end-survey-multiple-select.json",
          "version": 1.2,
          "compatibleWith": 1,
          "dataKey": "manual/trip_user_input",
          "labelVars": {
            "modes": {
              "key": "travel_mode",
              "type": "length"
            },
            "purposes": {
              "key": "destination_purpose",
              "type": "length"
            }
          },
          "labelTemplate": {
            "en": "${purposes} purpose${purposes > 1 ? 's': ''}, ${modes} mode${modes > 1 ? 's': ''}",
            "es": "${purposes} propósito${purposes > 1 ? 's': ''}, ${modes} modo${modes > 1 ? 's': ''}"
          }
        },
```

shankari commented:
> We want this approach to be better for i18n. We should be able to provide versions of labelTemplate in different languages, so I think it should be structured like this:

@JGreenlee, I agree! Note that with this change, we should be able to remove the `filled-in-label` section from the buttons since the filled in label will be the filled in label template.

```
"buttons": {
        "trip-notes": {
        },
...
          "filled-in-label": {
            "en": "",
            "es": ""
          }
}
```

sebastianbarry commented:
@shankari Jack and I have a question about the `js/enketo/survey/answer.js` for the `EnketoSurveyAnswer` factory. 
My notes for the commit we made was:
> answer.js
> - made it so that _lazyLoadConfig() loads the config into the local-to-answer.js-scoped variable "_config"
> - Made it so that all the places where _config are called are looking at the right spot (i.e. _config[surveyname].labelVars....)
>
> THIS is not the best way of doing this, a better solution would be to have the _lazyLoadConfig() function always return the config, and then rather than storing _config entirely in the EnketoSurveyAnswer factory, we only store the config within the scope of UseLabelTemplate (for example: useLabelTemplateVersionOfConfig = _lazyLoadConfig() ), but because _lazyLoadConfig() returns a promise this would require us to rewrite the funcitons in EnketoSurveyAnswer to accept this config as a promise

We were wondering what your thoughts are on this, and if you have any pointers to make the issue of `_lazyLoadConfig()`  returning a promise any easier, or if it's just better and quicker to go with the current way: having `_config` be a variable in the scope. This way it could be `undefined` at the point of accessing in some rare cases, or be modified as well in some cases

shankari commented:
@sebastianbarry I don't understand what the problem was that needed the fix in https://github.com/e-mission/e-mission-phone/pull/917#issuecomment-1379274701

sebastianbarry commented:
While the trip timestamps are being filled out locally, they are only being cached. This means that when we click refresh, the timestamp goes away again:



Jack mentioned that it is because the location that I am storing the `start_fmt_time` and `end_fmt_time` in does not get "saved". That saving happens somewhere in the `enketoSurvey` factory:
<img width="823" alt="image" src="https://user-images.githubusercontent.com/61334340/212422211-001c29e8-1de1-4e91-97d9-547f97ce88d2.png">


Keeping this in mind for when I make this fix***

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+                                            // for places, it is enter and exit time (end of one trip and start of next)
+
+    // Prefill the time bounds of the trip into the survey
+    const startDayAndTime = timebounds.start_fmt_time.split('T');
+    const startHMS = startDayAndTime[1].substring(0, 8); // truncated to 8 chars for HH:MM:SS
+    const startTimezoneCode = startDayAndTime[1].split('-')[1];
+    const startTime = `${startHMS}.000-${startTimezoneCode}`
+    const endDayAndTime = timebounds.end_fmt_time.split('T');
+    const endHMS = endDayAndTime[1].substring(0, 8); // truncated to 8 chars for HH:MM:SS
+    const endTimezoneCode = endDayAndTime[1].split('-')[1];
+    const endTime = `${endHMS}.000-${endTimezoneCode}`
```
at the very least, we should change this to use moment instead of this hardcoded parsing. But let's see if we can do anything more elegant as well.

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+          <start>${timebounds.start_fmt_time}</start>
+          <end>${timebounds.end_fmt_time}</end>
+          <group_hg4zz25>
+            <Date>${startDayAndTime[0]}</Date>    ${/* YY:MM:DD */''}
+            <Start_time>${startTime}</Start_time> ${/* HH:MM:SS.mmm-HH:MM */''}
+            <End_time>${endTime}</End_time>
```
We basically have two options for this:
(1) we can put this into the survey config as well, or
(2) we can specify that all "notes" surveys must have these fields to work

Right now, I am tempted to go with (2) and defer (1) until we have more examples of notes surveys.
As an aside, is the date needed? wouldn't the date just be the date of the trip/place?
To answer my own question: what if somebody was at a place for multiple days? To pick a completely hypothetical example, if they were in COVID isolation, for example.

shankari replied:
we are going to have an xml path in the survey config
while loading the survey config, we will load the xml in addition to the json
we will parse the XML and find the tag named `<instance>`
and will use that to fill in the prefilled value

Concretely, the instance for the time use survey is as below which seems to have all the strings that we need.

```
      <instance>
        <a88RxBtE3jwSar3cwiZTdn id="a88RxBtE3jwSar3cwiZTdn">
          <start/>
          <end/>
          <group_hg4zz25>
            <Date/>
            <Start_time/>
            <End_time/>
            <Activity_Type/>
            <Personal_Care_activities/>
            <Employment_related_a_Education_activities/>
            <Domestic_activities/>
            <Recreation_and_leisure/>
            <Voluntary_work_and_care_activities/>
            <Other/>
          </group_hg4zz25>
          <meta>
            <instanceID/>
          </meta>
        </a88RxBtE3jwSar3cwiZTdn>
      </instance>
```

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+    console.log('About to launch survey ', surveyName);
+
+    let partialTimeUseResponse;
+    if (surveyName == 'TimeUseSurvey') {
+      partialTimeUseResponse = $scope.getPartialTimeUseResponse();
+    }
+
```
This needs to be more more generic, right? I thought we could specify any type of survey for the notes button.

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+      partialTimeUseResponse = $scope.getPartialTimeUseResponse();
+    }
+
+    if ($event.stopPropagation) $event.stopPropagation();
```
Please comment on why this is needed

JGreenlee replied:
Otherwise, the click bubbles through to the list item and the details page opens as well as the survey

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+        if (!result) {
+          return;
+        }
+        result.start_fmt_time = moment.unix(result.start_ts).format();
+        result.end_fmt_time = moment.unix(result.end_ts).format();
```
Do the`fmt_time` fields need to be in the trip `properties`, like the `ts` fields?
https://github.com/e-mission/e-mission-phone/pull/917/commits/2072a3f152b4466d2440653ba21bbc5d9943a056

JGreenlee replied:
@sebastianbarry

sebastianbarry replied:
These lines no longer exist and are handled in `enketo-notes-list` by https://github.com/sebastianbarry/e-mission-phone/commit/7c9ddf28ec6bab3c95fabcf1afdfafb93957ee6a

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+  /**
+   * Embed 'inputType' to the trip.
+   */
+   enbs.extractResult = (results) => EnketoSurveyAnswer.filterByNameAndVersion('TimeUseSurvey', results);
```
again, this should not be hard-coded, right?

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+   */
+   enbs.extractResult = (results) => EnketoSurveyAnswer.filterByNameAndVersion('TimeUseSurvey', results);
+
+   enbs.processManualInputs = function(manualResults, resultMap) {
+    if (manualResults.length > 1) {
+        Logger.displayError("Found "+manualResults.length+" results expected 1", manualResults);
+    } else {
+        console.log("ENKETO: processManualInputs with ", manualResults, " and ", resultMap);
+        const surveyResult = manualResults[0];
+        resultMap[enbs.SINGLE_KEY] = surveyResult;
+    }
+  }
```
Is everything below this just a direct copy-paste from the existing `survey/enketo` button?
Can you highlight the changes as comments to make it easier to review?

shankari commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,287 @@
+            trip.trip_addition = [];
+        }
+        trip.tripAddition = [];
+        enbs.populateManualInputs(trip, trip.nextTrip, enbs.SINGLE_KEY,
+            manualResultMap[enbs.SINGLE_KEY]);
```
Note that a lot of this code can be simplified since there is only a SINGLE_KEY version of the trip additions - there is no MULTILABEL option. However, that is beyond the scope of this change.

shankari reviewed (changes requested):
@sebastianbarry @JGreenlee This looks very good! I have a few comments inline.

***Please resolve them until I merge this set of changes before starting on the place code.***

- if we want to use the trip code as a template for the place code, the trip code had better be in good shape
- it will make it much easier for me to review as well. This is already a pretty large PR, with 26 files, and adding the place code will just add to the complexity.

JGreenlee commented on www/css/main.diary.css:
```diff
@@ -216,6 +216,22 @@ a.item-content {
+  flex-wrap: wrap;
+}
+
+.diary-card.short {
```
I was using this for place list items because they don't need to be as tall as trips.
I also though they should have a different color to differentiate from trips, and have the content centered like the Sydney team's wireframes showed

shankari replied:
Makes sense. Can you add comments to the css for the locations where the change was not obvious?
It will help the next person who needs to work on the code...

JGreenlee commented on www/css/main.diary.css:
```diff
@@ -351,6 +371,10 @@ a.item-content {
   height: 50%;
 }
 
+.diary-infos:nth-child(2) {
+  justify-content: center;
+}
+
```
`nth-child` is a css feature to style the nth sibling. In this case, it's the 
bottom half of the diary card which contains the button. To prevent text overlapping we needed to center it

The root problem of all this is that the UI relies on fixed pixel values for heights and sizes, and is not structured for responsive design. We cannot fix that without an overhaul, so this remedy tries to make things at least look presentable

shankari replied:
I did not know this! The styling aspects of HTML are the ones that I am least familiar with since I am not very good at visual design and have generally handed that off to others for the entire duration of the project.

JGreenlee commented on www/js/config/dynamic_config.js:
```diff
@@ -25,14 +25,14 @@ angular.module('emission.config.dynamic', ['emission.plugin.logger'])
     dc.configReady = function() {
         if (dc.isConfigReady) {
-            Logger.log("UI_CONFIG in configReady function, resolving immediately");
+            // Logger.log("UI_CONFIG in configReady function, resolving immediately");
```
Strangely, these log statements cause errors for me that break the entire app

I commented them out temporarily and meant to ask you - but it looks like I forgot and committed them

shankari replied:
So these are calls to the native layers.
The native functions are only available after the native plugins are initialized.
So in the code, you may see `ionicPlatform.ready()` in many places.
That callback is invoked after the native layer is ready and the rest of the our code should be initialized from that callback.

So if you were making calls to the `DynamicConfig` outside of `ionicPlatform.ready()` you may run into this.

JGreenlee commented on www/js/diary/infinite_scroll_list.js:
```diff
@@ -29,41 +31,64 @@ angular.module('emission.main.diary.infscrolllist',['ui-leaflet',
+    if (trip.tripAddition) {
+      height += 40 * trip.tripAddition.length;
+    }
+    if ($scope.showPlaces) {
+      height += 120;
```
Currently, each item of the `collection-repeat` includes a trip and a place.

Once we are actually interweaving place entries with trip entries, I can change how this behaves so that entries are separate collection items

JGreenlee commented on www/js/diary/infinite_scroll_list.js:
```diff
@@ -190,6 +216,26 @@ angular.module('emission.main.diary.infscrolllist',['ui-leaflet',
         }
     });
+    Timeline.getUnprocessedLabels($scope.enbs).then(([pipelineRange, manualResultMap]) => {
+      if (pipelineRange.end_ts) {
+          $scope.$apply(() => {
+            $scope.data.manualResultMap = manualResultMap;
+          });
+          console.log("After reading in the label controller, manualResultMap "+JSON.stringify($scope.manualResultMap), $scope.data.manualResultMap);
+          $scope.infScrollControl.pipelineRange = pipelineRange;
+          $scope.infScrollControl.currentEnd = pipelineRange.end_ts;
+          $scope.infScrollControl.callback = function() {
+            $ionicScrollDelegate.scrollBottom(); // scrollTop()?
+          };
+          $scope.readDataFromServer();
+      } else {
+          $scope.$apply(() => {
+              $scope.infScrollControl.reachedEnd = true;
+          });
+          $ionicLoading.hide();
+          $scope.$broadcast('scroll.infiniteScrollComplete')
+      }
```
Agreed. I think @sebastianbarry and I talked about refactoring this

shankari replied:
This has been refactored. I have comments on the refactoring, but we can deal with them there.

JGreenlee commented on www/js/diary/infinite_scroll_place_item.js:
```diff
@@ -0,0 +1,47 @@
+angular.module('emission.main.diary.infscrollplaceitem',
+    ['emission.main.diary.infscrolllist',
+        'emission.survey.multilabel.services',
+        'emission.main.diary.infscrolldetail',
+        'ng-walkthrough',
+        'nvd3',
+        'emission.plugin.kvstore',
```
Agreed. I was not yet sure what they all do, so I included the ones I was unsure about to be safe
We should find which ones can be safely removed

sebastianbarry replied:
Confirmed these are not needed for the module and removed them: https://github.com/e-mission/e-mission-phone/pull/917/commits/269b565b1989760e0bd41461eca8ca445456e3ad

JGreenlee commented on www/js/diary/infinite_scroll_place_item.js:
```diff
@@ -0,0 +1,47 @@
+        start_fmt_time: $scope.trip?.end_fmt_time,
+        end_fmt_time: $scope.trip?.nextTrip?.start_fmt_time
+      }
+      if(!obj.end_fmt_time) obj.end_fmt_time = obj.start_fmt_time;
```
@sebastianbarry 

this is what prefills in the time use survey for that most recent place. We thought this was better than leaving it blank.
However, a better solution might be to grab the current time (because the last place is presumably your current location)

sebastianbarry replied:
Changed this to prefill with the current time as a default option (i.e. the last place in the list):

https://github.com/e-mission/e-mission-phone/pull/917/commits/82910086ea6ce874ff571abd79c5d049f38ab456

sebastianbarry commented:
WRT the `places-notes` list, I was able to get it to display the results of the placesAdditions notes objects in the places object, now I just need to get the display height to respond accordingly:

<img width="1302" alt="Screen Shot 2023-01-17 at 4 19 27 PM" src="https://user-images.githubusercontent.com/61334340/213033932-45c4b9dd-c375-4436-ae2e-d774bf7d5f19.png">

I found that when changing the values of `height` in this div that sits just within the `diary-card short` div, we can adjust the height of the entire card and get it to display how we want.

Without having max-height or min-height defined, this is what it looks like:
<img width="1294" alt="image" src="https://user-images.githubusercontent.com/61334340/213034263-b887c8f5-0d46-4a5a-b843-dd2d41d67bc2.png">

Which seems to have just a little bit too much dead space towards the bottom; This may just be a padding issue which I will investigate tomorrow

JGreenlee commented on www/css/main.diary.css:
```diff
@@ -337,6 +353,10 @@ a.item-content {
   /* height must be specified in the html */
 }
 
+.enhanced-trip-item .diary-map {
+  border-radius: 30px 0px 30px 0px;
+}
+
```
Yes, the root element of trip item has `enhanced-trip-item` class if notes are enabled. This changes the appearance of the map

JGreenlee commented on www/js/diary/infinite_scroll_trip_item.js:
```diff
@@ -3,24 +3,28 @@
-                                                        'emission.stats.clientstats',
-                                                        'emission.incident.posttrip.manual'
-])
+angular.module('emission.main.diary.infscrolltripitem',
+    ['emission.main.diary.infscrolllist',
```
It was intentional. I keep whitespace changes to a minimum, but when I notice things are messy and not matching the format elsewhere, I have been slightly tweaking
Do you want me to stop doing that?

shankari replied:
no that's fine. I have had other students whose editors made a bunch of unintentional whitespace changes, so I've become a bit paranoid about it.

JGreenlee commented on www/js/survey/enketo/answer.js:
```diff
@@ -35,23 +36,37 @@ angular.module('emission.survey.enketo.answer', [
+          const fieldName = config.labelVars[lblVar].key;
+          let fieldStr = _getAnswerByTagName(xmlDoc, fieldName);
+          if (fieldStr == '<null>') fieldStr = null;
+          if (config.labelVars[lblVar].type == 'length') {
+            const fieldMatches = fieldStr?.split(' ');
+            labelVars[lblVar] = fieldMatches?.length || 0;
+          } else {
+            labelVars[lblVar] = fieldStr || '';
+          }
```
Not yet. we left it open for other options but our use cases right now are just 'length'

I think we could implement a 'list', which puts the specific choices, not categories, in a comma separated list
However, I don't think we can use MessageFormat to format a list. [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat) will do it

shankari replied:
I think that a comma separated list will be too verbose. If we do want to support a `summary` option, I think it would be `first item (in the current language) +(length-1)` e.g. `(eating +2)` or `(சாப்பாடு + 2)`

sebastianbarry replied:
I am trying to implement this but I have a few questions about the `summary` option:

1. What fields do we want to be summarized? We can either do:
| If the results were `Domestic activities: 1 response (cleaning), Personal Care: 2 responses (doing_sport & sleeping), Recreation activities: 0 responses, etc. |`
| :-----: |
| "Domestic Activities + 1" |
| **OR** "doing_sport + 1" |

2. How do we want the labelTemplate to look? My thoughts:

For "length", we currently interpolate labelVars onto labelTemplate:
```
labelVars = {ral: 0, da: 1, pca: 2}
labelTemplate = { pca, plural, =0 {} other {# Personal Care, }}{ da, plural, =0 {} other {# Domestic, } }{ ral, plural, =0 {} other {# Recreation/Leisure, } }
```

Gives: `"2 Personal Care, 1 Domestic Activities"`

Now _(if we want to display like "Domestic Activities + 1"),_ I believe we need to change labelVars to keep track of how many responses each question has (`all: 3`), but labelTemplate must be changed somehow to display this correctly, my first idea is nesting:
```
labelVars = {all: 3, ral: 0, da: 1, pca: 2}
labelTemplate = { pca, plural, =0 { da, plural, =0 { ral, plural, =0 { No responses } other { Recreation and Leisure } } other { Domestic Activities + {all}-1 } } other { Personal Care + {all}-1 }}
```

shankari replied:
**I don't think we should implement the summary option right now**. We should wait until we have an actual use case of somebody who wants a different type of label and implement it then.

But if/when we decide to implement it, I think we would want `"doing_sport + 1"`

shankari replied:
for the record, the survey spec does in fact have multiple languages so the label is `How_old_are_you:label` without any ? etc. the spanish version is ¿Cuántos años tienes? and the english version is How old are you? So it seems like we should be able to find the string to display from the result (and presumably the result has the label `How_old_are_you` instead of any particular formatted language).

```
<fieldset class=\"question simple-select \"><fieldset><legend><span lang=\"es\" class=\"question-label \" data-itext-id=\"/aSfdnWs9LE6q8YEF7u9n85/group_hg4zz25/How_old_are_you:label\">¿Cuántos años tienes?</span><span lang=\"en\" class=\"question-label active\" data-itext-id=\"/aSfdnWs9LE6q8YEF7u9n85/group_hg4zz25/How_old_are_you:label\">How old are you?</span>
```

JGreenlee commented on www/js/survey/enketo/answer.js:
```diff
@@ -106,10 +122,12 @@ angular.module('emission.survey.enketo.answer', [
    */
   function resolveLabel(name, xmlDoc) {
-    return LABEL_FUNCTIONS[name](xmlDoc);
+    if (LABEL_FUNCTIONS[name])
+      return LABEL_FUNCTIONS[name](xmlDoc);
```
My understanding was that later on, we will write dedicated label functions for clients that ask for them, when the template is not enough

sebastianbarry replied:
Agreed. This conversation is good to resolve

shankari replied:
Right but there is no comment here indicating that. Please add the comment; then I can resolve this.

JGreenlee replied:
done bce2beb40365270aa931da237568a439fe96e3dd

JGreenlee commented on www/js/survey/input-matcher.js:
```diff
@@ -116,5 +116,37 @@ angular.module('emission.survey.inputmatcher', ['emission.plugin.logger'])
+    }
+
+    // remove deleted additions
+    tripAdditionList.forEach(addition => {
+        if (addition.data.status == "DELETED") {
+            // find and remove the original addition
+            const i = tripAdditionList.findIndex(a =>
+                a.data.end_ts == addition.data.end_ts
+                && a.data.start_ts == addition.data.start_ts);
+            tripAdditionList.splice(i, 1)
+            // also remove this deletion signifier
+            tripAdditionList.splice(tripAdditionList.indexOf(addition), 1);
+        }
+    });
```
I remember.

Whichever approach is taken, doesn't the filtering need to also happen on the phone for unprocessed additions? Here I am using start_ts and end_ts

shankari replied:
Correct, the filtering does need to happen on the phone for unprocessed additions. I just think that the matching code on both the phone and the server needs to be consistent.

shankari replied:
- if we do use id, we need a separate id since unprocessed additions will not have the mongodb id

so let's think about pros and cons briefly:
- advantage of ids is that we can match even if the start and end ts are different (e.g. if we support edits in the future)
- advantage of `start_ts` and `end_ts` is that we can be consistent with the other matchings

So if we are going to support edits in which people can change their time range on edit, then we must use ids. This will also be very complicated, so I don't really want to do it.

JGreenlee commented on www/js/survey/input-matcher.js:
```diff
@@ -116,5 +116,37 @@ angular.module('emission.survey.inputmatcher', ['emission.plugin.logger'])
+        }
+    });
+
+    // filter out trip additions that do not start withing the bounds of the trip
+    const matchingAdditions = tripAdditionList.filter((additionCandidate) =>
+        additionCandidate.data.start_ts >= trip.start_ts &&
+        additionCandidate.data.start_ts <= trip.end_ts);
```
Perhaps naively, I thought we could avoid the complexity of the other matching function. Maybe that is because I don't fully understand how drafts work. Why would an addition's time bounds not fall within the trip's time bounds?

shankari replied:
one of the issues we had reported by people earlier is that unprocessed trips wouldn't show up in the app. So if you took a trip, at the end of the trip, you would still see a "blank diary". So people didn't know if the app was working.

to address this, we display draft aka "green background" entries for unprocessed trips which we kind of hack together by looking at the transitions of the trip start and trip end from the data collection FSM. These are currently displayed in the diary, but not in the label screen.

so a concrete example would be:
- I just got home, 5 mins later, the FSM in the app has detected that the trip ended, and has attempted to push the trip data to the server
   - if it was unable to push (due to lack of network), the trip would be draft
   - if it was pushed, but the pipeline has not yet run, the trip would be draft

so note that the FSM on the phone will always have a bit of a lag for both trip start and trip end.
- On trip start, we need to exit the geofence or be moving for some substantive time before we start tracking, otherwise, we will start tracking everytime somebody goes to the kitchen of their house.
- On trip end, we need to wait until the user has been stationary for a while before we detect a trip end and stop tracking, otherwise we will stop the trip at every red light.

So the transitions from the phone FSM will lag the actual start and end in general.

In the processing, as part of the trip segmentation and clean and resample stages we attempt to fill these gaps. 

- Because we are post-processing, we know the end place and we can work backwards and find when we entered the end place and pick that as the trip end.
- Similarly, at the trip start, we know the last end place, so at trip start, we can "join" back to the previous place and extrapolate the time required to do so, and pick that as trip start.

JGreenlee commented on www/js/survey/enketo/service.js:
```diff
@@ -119,16 +110,8 @@ angular.module('emission.survey.enketo.service', [
-    } else {
-        return null;
-    }
+    const answer = _state.opts.prefilledSurveyResponse || _state.opts.trip?.userInput["SURVEY"];
```
`prefilledSurveyResponse` is only supplied for enketo-demographics and add-note-button
`userInput['SURVEY']` is only used for trip confirm / multilabel

We don't have a use case where both are defined. However in the case we did, I think it makes sense to prefer the prefilledSurveyResponse because it's been specifically defined

shankari replied:
Now that we want to support edits, I believe we will in fact have a use case in which both are defined. And in general, we should prioritize the prior user input over the partially filled template

JGreenlee commented on www/templates/survey/enketo/add-note-button.html:
```diff
@@ -0,0 +1,5 @@
+<div class="enketo-button">
+  <button class="button icon ion-plus button-primary" ng-click="openPopover($event, trip, 'NOTES')">
+    <b>{{displayLabel()}}</b>
```
Does Angularjs not have a way to define a property, cache it and then automatically recompute when its dependencies change?
I miss this from other frameworks

shankari replied:
It does do this, which is the `$digest`/`$apply` mechanism.
So if you define a variable in `$scope`, the value is cached unless you call a `$scope.$apply`, which will basically recompute the cached values using some kind of dependency graph.

But the point I was making here is that you are not using a variable, you are using a function. So while rendering, angular will *always* call the function since the function may return different results at different times. You want to replace all functions in the HTML (at least to the extent possible) with variables.

e.g. https://github.com/e-mission/e-mission-phone/commit/30561b94d8691e37fc3e50b8488f5da82f6dd815

JGreenlee replied:
6888ec5c8e4594cc250eea791a669fa4c59131b5 handles this

JGreenlee commented on www/templates/survey/enketo/notes_list.html:
```diff
@@ -0,0 +1,11 @@
+        <ul>
+            <li class="notes-list-entry" ng-repeat="entry in entries">
+                <b class="two-lines">{{entry.data.label}}</b>
+                <i>{{entry.data.start_fmt_time.substring(11,16) + ' - ' + entry.data.end_fmt_time.substring(11,16)}}</i>
```
I need to learn about Moment. I'll check it today
Sebastian told me about it some time after I wrote these

shankari commented on www/js/diary/infinite_scroll_detail.js:
```diff
@@ -11,12 +11,17 @@ angular.module('emission.main.diary.infscrolldetail',['ui-leaflet', 'ng-walkthro
-  $scope.surveyOpt = SurveyOptions.MULTILABEL;
-  $scope.tripFilterFactory = $injector.get($scope.surveyOpt.filter);
-  $scope.filterInputs = $scope.tripFilterFactory.configuredFilters;
+
+  DynamicConfig.configReady().then((configObj) => {
+    const surveyOptKey = configObj.survey_info['trip-labels'];
+    $scope.surveyOpt = SurveyOptions[surveyOptKey];
+    console.log('surveyOpt in infinite_scroll_details.js is', $scope.surveyOpt);
+    $scope.tripFilterFactory = $injector.get($scope.surveyOpt.filter);
+    $scope.filterInputs = $scope.tripFilterFactory.configuredFilters;
+  });
```
e.g. this is outside `ionicPlatform.ready()`

shankari commented:
For the demo today:
- (JG): see why just-added entries (without a refresh) are not deleted
- (SB): the start and end timestamps are always the same for every entry
- (SB): look at the height
- (KS): use `match_id` instead of `_id` on the server

shankari commented on www/js/diary/detail.js:
```diff
@@ -11,10 +11,15 @@ angular.module('emission.main.diary.detail',['ui-leaflet', 'ng-walkthrough',
+                                        DynamicConfig, CommHelper, PostTripManualMarker, $translate) {
+  console.log("controller DiaryDetailCtrl called with params = "+JSON.stringify($stateParams));
+
+  DynamicConfig.configReady().then((configObj) => {
+    const surveyOptKey = configObj.survey_info['trip-labels'];
+    $scope.surveyOpt = SurveyOptions[surveyOptKey];
+    console.log('surveyOpt in details.js is', $scope.surveyOpt);
+  });
+
```
Resolving since this is low priority

shankari commented on www/js/survey/enketo/enketo-notes-list.js:
```diff
@@ -32,7 +32,7 @@ angular.module('emission.survey.enketo.notes-list', [])
       console.log("Deleting entry", entry);
 
-      const dataKey = entry.metadata.key;
+      const dataKey = entry.key || entry.metadata.key;
```
do we actually have to have the `dataKey` as a field for the notes list?
We already support different surveys and different labels.
And the data is going to come back as `trip-additions` (please see server code)
So it seems like we don't really need the data key to be configurable.

Keeping a single data key for "one per trip" and one for "n per trip" objects respectively can also make it easier for the downstream code, such as the public dashboard.

Again, check the server code for an example

JGreenlee replied:
I thought the key was necessary here because for deletion, we need to put an entry with status DELETED to the UserCache. To my understanding we are using a different key for trip- and place- level additions (`manual/trip_addition_input` and `manual/place_addition_input`)
so notes-list needs to know which key to use when it adds the DELETED entry.

shankari commented on www/js/diary/services.js:
```diff
@@ -412,10 +412,20 @@ angular.module('emission.main.diary.services', ['emission.plugin.logger',
-                manualFactory.processManualInputs(manualResults, manualConfirmResults);
-                return [result, manualConfirmResults];
+            const enbsConfirmResults = {};
+            return new Promise((rs, rj) => {
+                Promise.all(manualPromises).then(function(results) {
+                  manualFactory.processManualInputs(results, manualConfirmResults);
+                });
+                Promise.all(enbsPromises).then(function(results) {
+                  enbs.processManualInputs(results, enbsConfirmResults);
+                });
+                rs([result, manualConfirmResults, enbsConfirmResults]);
```
I'm not sure this actually works - in the old code, we would only return after all the `manualPromises` had completed (return is within the `then`, but here, we don't wait for `manualPromises` or `enbsPromises` before we call `rs`.

I think you need

```
return Promise.all(manualPromises.concat(enbsPromises)).then([comboResults] => {
      manualResults = comboResults.slice(0,manualPromises.length);
...
      return [result, manualConfirmResults, enbsConfirmResults];
});
```

JGreenlee replied:
You're absolutely right. I'll implement this with spread syntax

shankari commented on www/js/survey/enketo/service.js:
```diff
@@ -139,6 +139,12 @@ angular.module('emission.survey.enketo.service', [
+        // generate a UUID v4 for the survey response
+        // using method from https://stackoverflow.com/a/2117523
+        // crypto.generateUUID() is preferable, but not availble to us in the app
+        data.match_id = ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
+          (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
+        );
```
this is fine, although given that this doesn't need to be cryptographically secure, maybe we can use a the lighter weight objectID outlined here: objectID
https://stackoverflow.com/a/66332305/4040267

This is an optimization though so am happy with deferring it until later (if ever)
You can resolve this as soon as you have read it.

JGreenlee replied:
done in 586e1abc74ddbb700ed43460f2d816245f2f9b16

shankari reviewed (changes requested):
Comments on the changes from yesterday. We are getting there 😄

sebastianbarry commented:
I did get the time to display properly, but @JGreenlee and @shankari can you both take a look at how I implemented this in enketo-add-note-button.js? I do not know if there's a better way to go about this than hard-coding the survey ID and group name into setting the variable `start-time` and `end-time`:
```javascript
let start_time = result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Date + " " + result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Start_time.substr(0, 12);
        let end_time = result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Date + " " + result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.End_time.substr(0, 12)
        result.start_fmt_time = moment(start_time).format("LT")
        result.end_fmt_time = moment(end_time).format("LT")
```

sebastianbarry commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,311 @@
+        let start_time = result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Date + " " + result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Start_time.substr(0, 12);
+        let end_time = result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.Date + " " + result.jsonDocResponse.a88RxBtE3jwSar3cwiZTdn.group_hg4zz25.End_time.substr(0, 12)
+        result.start_fmt_time = moment(start_time).format("LT")
+//        moment.unix(result.start_ts).format();
```
Remove this comment
```suggestion
```

sebastianbarry commented on www/js/survey/enketo/enketo-add-note-button.js:
```diff
@@ -0,0 +1,311 @@
+        result.start_fmt_time = moment(start_time).format("LT")
+//        moment.unix(result.start_ts).format();
+        result.end_fmt_time = moment(end_time).format("LT")
+//        moment.unix(result.end_ts).format();
```
Remove this comment also
```suggestion
```

shankari commented:
> I do not know if there's a better way to go about this than hard-coding the survey ID and group name into setting the variable start-time and end-time:

@sebastianbarry `_getAnswerByTagName` is your friend. Note that we will need to have standard field names for the start and end timestamps anyway.
https://github.com/e-mission/e-mission-phone/pull/917#discussion_r1071783631

sebastianbarry commented on www/js/diary/infinite_scroll_place_item.js:
```diff
@@ -0,0 +1,47 @@
+        isPlace: true,
+        start_fmt_time: $scope.trip?.end_fmt_time,
+        end_fmt_time: $scope.trip?.nextTrip?.start_fmt_time
+      }
```
I found that for this change, `isPlace` is actually a useful variable https://github.com/e-mission/e-mission-phone/pull/917/commits/23be4a8b75b7d069a64b49add0c6be3077344058

shankari replied:
@sebastianbarry this is primarily polish at this point, so feel free to resolve this again, but theoretically, you don't need `isPlace`.

wrt "- split the one previous "start/stop time" lines into one that checks isplace, then uses enter/exit, otherwise uses start/stop", you can simply use enter/exit if they exist and start/stop if they don't

sebastianbarry replied:
Confirmed this was changed in commit https://github.com/sebastianbarry/e-mission-phone/commit/23be4a8b75b7d069a64b49add0c6be3077344058 to 
```javascript
        enter_fmt_time: $scope.trip?.end_fmt_time,
        exit_fmt_time: $scope.trip?.nextTrip?.start_fmt_time
      }
```

This is good to resolve

shankari replied:
when we switch to confirmed places, you should use the `enter_fmt_time` and `exit_fmt_time` from the place instead. you can resolve this comment after reading this message.

sebastianbarry commented on www/js/diary/infinite_scroll_trip_item.js:
```diff
@@ -31,14 +35,27 @@ angular.module('emission.main.diary.infscrolltripitem', [
+      return SurveyOptions[surveyOptKey];
+    }
+    
+    $scope.timeBounds = () => {
+      return {
+        isPlace: false,
+        start_fmt_time: $scope.trip?.start_fmt_time,
+        end_fmt_time: $scope.trip?.end_fmt_time
+      };
+    };
```
Added comments here: https://github.com/e-mission/e-mission-phone/pull/917/commits/c9b6f3dbbdb2c7f48f2b155fac482f6188e33b7d

sebastianbarry commented on www/js/diary/infinite_scroll_list.js:
```diff
@@ -29,41 +31,66 @@ angular.module('emission.main.diary.infscrolllist',['ui-leaflet',
   // TODO: load only a subset of entries instead of everything
 
   console.log("controller InfiniteDiaryListCtrl called");
+  
```
```suggestion
```
Remove extraneous whitespace

sebastianbarry commented on www/templates/diary/infinite_scroll_list.html:
```diff
@@ -34,13 +34,14 @@
 		<div collection-repeat="trip in data.displayTrips" item-height="getTripHeight(trip)">
             <div>
-                
+
```
Resolved: https://github.com/e-mission/e-mission-phone/pull/917/commits/ca18f97ff1613e2234b482c24936da1561c2e9ef

shankari commented:
Changes from today look pretty good in general. Main high level comments are:
-  it looks like you are doubling down on the hardcoding of the survey formats:
    - there are now a separate hardcoded place survey and hardcoded trip survey
    - neither of them use the instance from the survey json as outlined in the pending comment
       https://github.com/e-mission/e-mission-phone/pull/917#discussion_r1073850058
    - the fmt_time parts have not taken https://github.com/e-mission/e-mission-phone/pull/917#issuecomment-1397876302 into account either

I assume you are going to remove all the hardcoded survey bits before my next review.

I think that there is too much repeated code with the place vs, trip distinction instead of checking the presence or absence of exit_ts and initializing fields accordingly.

But otherwise, we are getting there...

shankari commented:
Next high level question is: how do we generalize the `EnketoNotesButtonService`?
Concretely, how do we pass in the `key` and the `survey name` given that we have separate keys for trips and places and we might have different surveys in the future?

(1) we shouldn't overthink this because we do want to move away from angular1 later this year and other frameworks will not necessarily have the service vs. controller difference
(2) we will only have two services - trips and places, so in `EnketoNotesButtonService`, it seems like we can set `MANUAL_KEYS = ["manual/trip_addition_input", "manual/place_addition_input"]`. This will allow us to get the keys at the same time, which is what we want anyway.
for the survey name, it seems like we may have different survey names for trips and places, so I would assume we want to create an `init` function that will again support the survey name for trips and places.

To simplify this, it seems like we can do something like

```
enbs.initConfig(tripSurveyName, placeSurveyName) {
    enbs.MANUAL_KEYS = []
    enbs.tripSurveyName = tripSurveyName;
    if (tripSurveyName) {
       enbs.MANUAL_KEYS.push("manual/trip_addition_input")
    }
    enbs.placeSurveyName = placeSurveyName;
    if (placeSurveyName) {
       enbs.MANUAL_KEYS.push("manual/place_addition_input")
    }
}

enbs.extractResult = function(results) {
    tripResults = EnketoSurveyAnswer.filterByNameAndVersion(enbs.tripSurveyName, results);
    placeResults = EnketoSurveyAnswer.filterByNameAndVersion(enbs.tripSurveyName, results);
    return tripResults.concat(placeResults);
   // or if it is important to maintain the separation for some reason
   // return [tripResults, placeResults];
}
```

JGreenlee commented:
The above 3 commits address the timestamp issues which have been plaguing us

The remaining issue which I described in 911c74c is that the survey will not allow us to display times in a the timezone of the trip (if it differs from local timezone). Enketo will always override our prefilled time and convert to the user's current local time, and I don't think there is a solution to this except modifying Enketo. We could submit a PR to them - not sure if it is a high enough priority for us right now

JGreenlee commented:
@shankari I know that trips use 'start' and 'end' terminology, while places use 'enter' and 'exit'. This is logical.
But we're not going to use 'enter' and 'exit' for additions, right? If I add an activity at a place, that activity has a 'start' and 'end' time, not an 'enter' or 'exit' time

shankari commented:
@JGreenlee let's discuss https://github.com/e-mission/e-mission-phone/pull/917#issuecomment-1401079011 when we move to the place changes. My rationale for leaving it as enter vs. exit was that, for completeness, I wanted to have `user_input` and `addition_input` for places as well.

To elaborate, for both trip-like and place-like objects, we want to support a 1:1 input (`XXX_user_input`) and a 1:n input (`XXX_addition_input`). We don't currently have any use cases for the `place_user_input`, but building out that structure in the data model makes it easy to think about, and creates some strong scaffolding for the future.

JGreenlee commented:
We can now have the dynamic config checked for updates on launch!

`dev-emulator-program.nrel-op.json`
```json
{
    "version": 1,
    "autoRefresh": true,
    ...
}
```
If the above config is loaded into the app (from github) and then it is incremented to version 2, the app will load the new version on its next launch. If `autoRefresh` is not given, the default behavior is the same.
This will be helpful for development, and potentially for deployers as well

We should now update the docs to explain this
@sebastianbarry Do you want to update the readme you made? Or if you'd rather give me write access to sebastianbarry/e-mission-docs I can do it later

JGreenlee commented on www/js/config/dynamic_config.js:
```diff
@@ -138,29 +169,25 @@ angular.module('emission.config.dynamic', ['emission.plugin.logger'])
+            }
+            // if 'autoRefresh' is set, we will check for updates
+            if (existingConfig.autoRefresh) {
+                loadNewConfig(existingConfig.joined, false, existingConfig.version).then((wasUpdated) => {
```
A 'reload' from devtools does the trick for me, or triggering an autoreload from phonegap by saving any file.
To me, this is faster than navigating to the Developer zone would be

When the time comes, it will be easy to change this from a dev feature (wait for the new config to be loaded before use) to a production feature (use the old config this time, but keep checking in background and let updates apply next time) - since it is already written in an asynchronous way

sebastianbarry commented on www/js/metrics-mappings.js:
```diff
@@ -374,12 +374,16 @@ angular.module('emission.main.metrics.mappings', ['emission.plugin.logger',
             this.populateCustomFootprints();
         });
+      } catch (err) {
+        console.error("Error in Metrics Mappings init, " + err);
```
This was a change I made in commit https://github.com/e-mission/e-mission-phone/commit/d2ddae8d9f5de413b0282269482895a20d57bead

In the commit, I reasoned it to "there was a single "undefined" error causing the profile screen to not load"

I undid this commit and reloaded the app and it is working fine without the try/catch. I think we are good to not have this (essentially I think this was leftover code from debugging that made it into the PR)

@shankari do you think this try/catch is still useful to have? I do not believe so

JGreenlee replied:
I remember that bug. it was probably fixed in c39018a

If we can verify that, I see no reason to keep the try/catch

sebastianbarry replied:
I confirmed that commit fixed the bug, as it did not show up after that commit.

I pushed https://github.com/e-mission/e-mission-phone/pull/917/commits/1d2725301ecc7611373e33900ce5f54a9f264218 this is good to resolve

shankari commented on www/js/diary/infinite_scroll_place_item.js:
```diff
@@ -0,0 +1,45 @@
+    // this allows us to pre-fill time and date in surveys that have 'Date', 'Start_time', and 'End_time' fields
+    $scope.timeBounds = () => {
+      let obj = {
+        isPlace: true,
+        enter_fmt_time: $scope.trip?.end_fmt_time,
+        exit_fmt_time: $scope.trip?.nextTrip?.start_fmt_time
```
once we start supporting places, we can and should use the place `enter` and `exit` time directly instead of inferring from the trip.

shankari commented on www/js/survey/enketo/answer.js:
```diff
@@ -35,23 +36,37 @@ angular.module('emission.survey.enketo.answer', [
+          if (config.labelVars[lblVar].type == 'length') {
+            const fieldMatches = fieldStr?.split(' ');
+            labelVars[lblVar] = fieldMatches?.length || 0;
+          } else {
+            labelVars[lblVar] = fieldStr || '';
+          }
```
I would throw an exception here instead of returning the fieldStr. Code that is not tested might as well not be there.
And we are not planning to support anything other than length right now.

sebastianbarry replied:
I did this in this commit: https://github.com/e-mission/e-mission-phone/pull/917/commits/9cab667c6503e68edc2d2b49a8b2937b4716f43a

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/61334340/216093191-c994dbf5-697c-4443-afcd-35008eb66e09.png">

shankari commented on www/js/survey/enketo/answer.js:
```diff
@@ -106,14 +122,40 @@ angular.module('emission.survey.enketo.answer', [
+    // return unix timestamps in milliseconds
+    return {
+      start_ts: momentStart.unix(),
+      end_ts: +momentEnd.unix()
```
nit: why is there a + before the `momentEnd`?

sebastianbarry replied:
```suggestion
      end_ts: momentEnd.unix()
```
I tested this works without the "+". It looks like a typo to me

shankari commented on www/templates/diary/trip_list_item.html:
```diff
@@ -1,105 +1,98 @@
-                    e.g. https://github.com/e-mission/e-mission-docs/issues/674#issuecomment-933083710
-                    -->
-                    <div class="col-90">
+                <div class="diary-infos" style="padding-top: 10px; padding-bottom: 10px; width: 100%;">
+                    <!-- Mode and Purpose Selection -->
+                    <div class="row" style="margin-top: 0px; padding: 4px">
+                        <!-- Assumes that the trip labels have already been
+                        filled in outside the directive to filter properly
+                        e.g. https://github.com/e-mission/e-mission-docs/issues/674#issuecomment-933083710
+                        -->
+                        <div class="col-90">
```
Extraneous whitespace?

JGreenlee replied:
The formatting of the HTML was quite poor before this change. Do you want to go back to how it was?

shankari replied:
I don't want to spend too much time on this, but in what way was the formatting worse? This seems like it basically just moved everything in by 4 spaces/1 tab.

JGreenlee replied:
some start and close tags did not align, some children elements were not indented, and some `>` were hanging on their own line

before
https://github.com/sebastianbarry/e-mission-phone/blob/9dafcc21f303e60a94250d1c9c608939b2637486/www/templates/diary/trip_list_item.html

after
https://github.com/sebastianbarry/e-mission-phone/blob/6f2100e4e67d092b1d35da4b4d5631f787d4032c/www/templates/diary/trip_list_item.html

shankari commented on www/templates/diary/trip_list_item.html:
```diff
@@ -1,105 +1,98 @@
                     </div>
-                </div>
 
-                <!-- Modes / Percentages -->
-                <div class="diary-modes-percents">
-                    <span
-                        class="diary-modes-percents-text"
-                        ng-if="!trip.isDraft"
-                        ng-repeat="sectionPct in trip.percentages"
-                    >
-                        <span class="diary-modes-percents-icon {{sectionPct.icon}}"></span>
-                        <span>{{sectionPct.pct}}%</span>
-                    </span>
-                    <button
-                    class="diary-draft-button"
-                    ng-if="trip.isDraft"
-                    ng-click="explainDraft($event)"
-                  >
-                    <span translate>{{'.draft'}}</span>
-                  </button>
-                </div>
```
we will likely restore these when we unify the diary and label views. So we might want to leave them in there so we know where they are coming from

JGreenlee replied:
These Modes / Percentages elements have not been removed - this change just fixed the formatting as the indentation of the entire file was wrong

shankari reviewed (changes requested):
This is looking pretty good to me so far. The two issue that I see are:
- the matching code is not correct
- the HTML file still has a lot of style-based function calls that should be replaced with variables instead.

shankari commented on www/js/survey/enketo/service.js:
```diff
@@ -100,18 +100,38 @@ angular.module('emission.survey.enketo.service', [
-    const answer = _state.opts.prefilledSurveyResponse || _state.opts.trip?.userInput["SURVEY"];
-    return answer ? answer.data.xmlResponse : null;
+    const answer = _state.opts.trip?.userInput["SURVEY"] || _state.opts.prefilledSurveyResponse;
+    return answer?.data?.xmlResponse || _getPrefilledModel(_state.opts.prefillFields) || null;
```
I think this has actually been changed in a subsequent commit

Yup, it is now

```
    const answer = _state.opts.prefilledSurveyResponse;
    return answer || _getPrefilledModel(_state.opts.prefillFields) || null;
```

JGreenlee commented on www/js/diary/infinite_scroll_list.js:
```diff
@@ -229,6 +229,18 @@ angular.module('emission.main.diary.infscrolllist',['ui-leaflet',
+      // TODO support places
+      const matchingTimelineEntry = $scope.data.displayTrips.find((trip) => {
+        return addition.data.start_ts >= ~~trip.start_ts && addition.data.start_ts <= trip.end_ts
+      });
```
done in 4c3f66454440411a4e422754d94abe4d0c6d6758

JGreenlee commented on www/js/survey/input-matcher.js:
```diff
@@ -117,29 +117,31 @@ angular.module('emission.survey.inputmatcher', ['emission.plugin.logger'])
+    // and filter out additions that do not start within the bounds of the trip
+    const matchingAdditions = im.getNotDeletedCandidates(tripAdditionList).filter((additionCandidate) =>
+        additionCandidate.data.start_ts >= ~~trip.start_ts &&
+        additionCandidate.data.start_ts <= ~~trip.end_ts);
```
done in f121498e956a719847b14fc65f465193460ecea4

shankari commented on www/js/diary/infinite_scroll_list.js:
```diff
@@ -232,9 +235,9 @@ angular.module('emission.main.diary.infscrolllist',['ui-leaflet',
-      });
+      const matchingTimelineEntry = $scope.data.displayTrips.find((trip) => 
+        InputMatcher.validUserInputForTrip(trip, addition, false)
+      );
```
Yes. This is what I meant. No need to copy and paste code and come up with duplicate and slightly different implementations.

shankari commented on www/js/survey/enketo/answer.js:
```diff
@@ -133,23 +133,30 @@ angular.module('emission.survey.enketo.answer', [
+    let additionEndTs = moment(date + 'T' + end).unix();
 
-    // return unix timestamps in milliseconds
+    // if the start and end timestamps are within the same minute, use the trip start and end timestamps
+    if (additionStartTs - (additionStartTs % 60) == trip.start_ts - (trip.start_ts % 60))
+      additionStartTs = trip.start_ts;
+    if (additionEndTs - (additionEndTs % 60) == trip.end_ts - (trip.end_ts % 60))
+      additionEndTs = trip.end_ts;
```
And what happens if we don't? Please update comment.

```
- the survey is only precise to the minute, while trips are precise to milliseconds
- if the start/end timestamps resolved from the survey response are within the same minute as the trip start/end, we will use the exact trip start/end
- this is to avoid addition timestamps being slightly off because of truncation issues
```

JGreenlee replied:
done a285285d6743087375b522366fabb3168a2fdfa5

shankari reviewed (changes requested):
Only high level issue that is left is to avoid using functions for classes/conditionals etc in the HTML

JGreenlee commented on www/templates/diary/place_list_item.html:
```diff
@@ -0,0 +1,23 @@
+                    <b>{{trip.end_display_name}}</b>
+                </div>
+                <div class="row" style="padding-top: 10px">
+                    <enketo-add-note-button notes-config="configPlaceNotes()" surveys="surveys" trip="trip"
```
done in 6888ec5c8e4594cc250eea791a669fa4c59131b5

JGreenlee commented on www/templates/diary/trip_list_item.html:
```diff
@@ -1,105 +1,98 @@
-<div>
+<div ng-class="{'enhanced-trip-item': configTripNotes()}">
```
done in 6888ec5c8e4594cc250eea791a669fa4c59131b5

shankari commented on www/js/diary/infinite_scroll_place_item.js:
```diff
@@ -41,5 +41,10 @@ angular.module('emission.main.diary.infscrollplaceitem',
-    $scope.configPlaceNotes = () => $scope.config?.survey_info?.buttons?.['place-notes'];
+
+    // config will initially be undefined, so we will watch
+    $scope.$watch('config', function (loadedConfig) {
+      // if place-notes button config is not present, then we won't even show the 'add note button'
+      $scope.configPlaceNotes = loadedConfig?.survey_info?.buttons?.['place-notes'];
+    });
```
nice!
For the record, I believe you could also handle this with a `$apply` in the config callback. This is effectively a config callback since the config is changed in the `$scope.$apply` of the config callback, right?

shankari reviewed (approved):
I don't see any pending issues for the trip code.
